Add AWS X-Ray propagator

shengxil opened the pull request:
Adds support for AWS X-Ray trace header propagation and tests.

anuraaga commented on sdk_contrib/aws_v1_support/src/main/java/io/opentelemetry/sdk/contrib/trace/aws/propagation/AWSXRayPropagator.java:
```diff
@@ -0,0 +1,238 @@
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.contrib.trace.aws.propagation;
```
Can you run `./gradlew googleJavaFormat`?

shengxil replied:
will do in rev2

anuraaga commented on sdk_contrib/aws_v1_support/src/main/java/io/opentelemetry/sdk/contrib/trace/aws/propagation/AWSXRayPropagator.java:
```diff
@@ -0,0 +1,238 @@
+    String parentId = spanContext.getSpanId().toLowerBase16();
+    char samplingFlag = spanContext.getTraceFlags().isSampled() ? IS_SAMPLED : NOT_SAMPLED;
+
+    String traceHeader = TRACE_ID_KEY + KV_DELIMITER + xrayTraceId + TRACE_HEADER_DELIMITER +
```
Can we add `tracestate` as additional params too? Not sure we'd prefer `tracestate=` some serialized string or put the KV's directly (probably need to dedupe against our known keys).

If not now, let's add a TODO to revisit it we shouldn't just drop it if possible.

shengxil replied:
I'll add a TODO here since we are discussing whether we will put the OT first 8 digit in trace state. I'd do them together.

anuraaga commented on sdk_contrib/aws_v1_support/src/main/java/io/opentelemetry/sdk/contrib/trace/aws/propagation/AWSXRayPropagator.java:
```diff
@@ -0,0 +1,238 @@
+
+  @Override
+  public <C> Context extract(Context context, C carrier, Getter<C> getter) {
+    checkNotNull(carrier, "carrier");
+    checkNotNull(getter, "getter");
```
```suggestion
    Objects.requireNonNull(carrier, "carrier");
    Objects.requireNonNull(getter, "getter");
```

shengxil replied:
will fix in rev2

anuraaga commented on sdk_contrib/aws_v1_support/src/main/java/io/opentelemetry/sdk/contrib/trace/aws/propagation/AWSXRayPropagator.java:
```diff
@@ -0,0 +1,238 @@
+
+  @Override
+  public <C> void inject(Context context, @Nullable C carrier, Setter<C> setter) {
+    Objects.requireNonNull(context, "context");
```
As per 3) should we avoid throwing on null?

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/error-handling.md#basic-error-handling-principles

shengxil replied:
I'm aligning with the [default propagator](https://github.com/open-telemetry/opentelemetry-java/blob/master/api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java#L80) and the official [b3](https://github.com/open-telemetry/opentelemetry-java/blob/4808387afbb8776669db67cc1b9c69824901fe90/contrib/trace_propagators/src/main/java/io/opentelemetry/contrib/trace/propagation/B3PropagatorInjectorSingleHeader.java#L42)/[jaeger](https://github.com/open-telemetry/opentelemetry-java/blob/master/contrib/trace_propagators/src/main/java/io/opentelemetry/contrib/trace/propagation/JaegerPropagator.java#L88) propagator. How about we keep the same behavior here and create a separate issue to change all propagators?

anuraaga commented on sdk_contrib/aws_v1_support/src/main/java/io/opentelemetry/sdk/contrib/trace/aws/propagation/AWSXRayPropagator.java:
```diff
@@ -0,0 +1,238 @@
+    Objects.requireNonNull(context, "context");
+    Objects.requireNonNull(setter, "setter");
+
+    Span span = TracingContextUtils.getSpanWithoutDefault(context);
```
Think it's better to use `.getSpan(context)` and check `span.getContext().isValid()`. From what I can tell, this current method may return an invalid span too, so either we check both null and valid or just valid.

shengxil replied:
will check valid to align with the [default propagator](https://github.com/open-telemetry/opentelemetry-java/blob/master/api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java#L83).

anuraaga commented on sdk_contrib/aws_v1_support/src/main/java/io/opentelemetry/sdk/contrib/trace/aws/propagation/AWSXRayPropagator.java:
```diff
@@ -0,0 +1,238 @@
+
+  @SuppressWarnings("StringSplitter")
+  private static <C> SpanContext getSpanContextFromHeader(C carrier, Getter<C> getter) {
+    try {
```
Instead of a large catch-all exception, only catch where an exception could happen. Even if we're worried about missing something, this isn't the right layer to have a catch-all.

anuraaga commented on sdk_contrib/aws_v1_support/src/main/java/io/opentelemetry/sdk/contrib/trace/aws/propagation/AWSXRayPropagator.java:
```diff
@@ -0,0 +1,238 @@
+  private static <C> SpanContext getSpanContextFromHeader(C carrier, Getter<C> getter) {
+    try {
+      String traceHeader = getter.get(carrier, TRACE_HEADER_KEY);
+      if (StringUtils.isNullOrEmpty(traceHeader)) {
```
I'm not sure if `internal` package is really meant for use from contrib, but similar to above, since this is SDK, we may as well just use Guava and `Strings.isNullOrEmpty`

shengxil replied:
Agree we can drop internal package usages. will fix in rev2

anuraaga commented on sdk_contrib/aws_v1_support/src/main/java/io/opentelemetry/sdk/contrib/trace/aws/propagation/AWSXRayPropagator.java:
```diff
@@ -0,0 +1,238 @@
+        return SpanContext.getInvalid();
+      }
+
+      String[] parts = traceHeader.split(String.valueOf(TRACE_HEADER_DELIMITER));
```
We want our instrumentation to have as low an impact on user app's performance as possible so we should take care in optimizing especially parse logic. 

This is probably the most optimized version

https://github.com/openzipkin/zipkin-aws/blob/master/brave-propagation-aws/src/main/java/brave/propagation/aws/AWSExtractor.java#L72

Even if that's too complex, we can still get a lot of improvement by avoiding `.split` in favor of `indexOf` and `substring`, I don't think we need to use `split` at all here (the warning is valid :) )

shengxil replied:
JDK has a specialized optimization on splitting by single character, which doesn't use regex. See [SO post](https://stackoverflow.com/questions/11001330/java-split-string-performances) and [JDK implementation](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/lang/String.java#l2324). Hence I don't think changing to the extractor or `substring` will make noticeable performance improvement that worths the trade of readability.

anuraaga replied:
It's not about the regex but the allocation of an array is still significant, especially since we don't need it. On the flip side, the change in readability also shouldn't be significant.

shengxil replied:
Ok that makes sense. Fixing in rev3

anuraaga commented on sdk_contrib/aws_v1_support/src/main/java/io/opentelemetry/sdk/contrib/trace/aws/propagation/AWSXRayPropagator.java:
```diff
@@ -0,0 +1,238 @@
+      TraceFlags traceFlags = null;
+
+      for (String part : parts) {
+        String[] keyAndValue = extractKeyAndValue(part);
```
When avoiding `split` we should also inline the key/value parse here to avoid allocating arrays

shengxil replied:
Same as above

anuraaga commented on sdk_contrib/aws_v1_support/src/main/java/io/opentelemetry/sdk/contrib/trace/aws/propagation/AWSXRayPropagator.java:
```diff
@@ -0,0 +1,238 @@
+            traceFlags = parseTraceFlag(value);
+            break;
+          default:
+            // Ignoring the arbitrary TraceHeader keys in OpenTelemetry
```
Think we can put them in tracestate, related to above comment. We should do both or neither though not just one.

shengxil replied:
I'll put a TODO for now

anuraaga commented on sdk_contrib/aws_v1_support/src/main/java/io/opentelemetry/sdk/contrib/trace/aws/propagation/AWSXRayPropagator.java:
```diff
@@ -0,0 +1,238 @@
+    if (xraySampledFlag.length() != SAMPLED_FLAG_LENGTH) {
+      throw new IllegalArgumentException("Invalid Sampling flag: " + xraySampledFlag);
+    }
+
```
```java
char flag = xraySampledFlag.charAt(0);
if (flag == IS_SAMPLED)
etc
```

shengxil replied:
Will fix in rev2. But I want to keep the length check since `Sampled=123` is invalid in X-Ray context.

anuraaga commented on sdk_contrib/aws_v1_support/src/test/java/io/opentelemetry/sdk/contrib/trace/aws/propagation/AWSXRayPropagatorTest.java:
```diff
@@ -0,0 +1,278 @@
+    assertThat(carrier)
+        .containsEntry(
+            TRACE_HEADER_KEY,
+            generateTraceHeaderValue(
```
I'd inline the strings, or have constants for the various forms. It's more readable and avoids the issue of now we need a test for this test helper.

shengxil replied:
Will fix in rev2.

anuraaga commented on sdk_contrib/aws_v1_support/src/test/java/io/opentelemetry/sdk/contrib/trace/aws/propagation/AWSXRayPropagatorTest.java:
```diff
@@ -0,0 +1,278 @@
+        null,
+        new Setter<Map<String, String>>() {
+          @Override
+          public void set(Map<String, String> ignored, String key, String value) {
```
This test doesn't seem to reflect reality and I don't think it actually verifies anything. We've already verified in our other tests that `carrier` is passed to the setter. So I think we can remove the test.

shengxil replied:
Will remove in rev2.

anuraaga commented on sdk_contrib/aws_v1_support/src/test/java/io/opentelemetry/sdk/contrib/trace/aws/propagation/AWSXRayPropagatorTest.java:
```diff
@@ -0,0 +1,278 @@
+  }
+
+  @Test
+  public void extract_SampledContext() {
```
Nit I'd put the success cases above the error cases.

shengxil replied:
Will fix in rev2.

anuraaga commented on sdk_contrib/aws_v1_support/src/test/java/io/opentelemetry/sdk/contrib/trace/aws/propagation/AWSXRayPropagatorTest.java:
```diff
@@ -0,0 +1,278 @@
+    Map<String, String> invalidHeaders = new LinkedHashMap<>();
+    invalidHeaders.put(
+        TRACE_HEADER_KEY,
+        generateTraceHeaderValue(
```
Here especially, don't use a helper method or even a constant, inlining the header will make much more readable what the test is checking.

shengxil replied:
will fix in rev2

anuraaga commented on sdk_contrib/aws_v1_support/src/test/java/io/opentelemetry/sdk/contrib/trace/aws/propagation/AWSXRayPropagatorTest.java:
```diff
@@ -0,0 +1,278 @@
+  @Test
+  public void extract_NotEnoughParts() {
+    Map<String, String> invalidHeaders = new LinkedHashMap<>();
+    invalidHeaders.put(TRACE_HEADER_KEY, "aa:bb:cc");
```
I think these two tests are supposed to use semicolon instead of colon.

shengxil replied:
The tests are actually invalid. They are copied from B3Propagator tests. I'll remove them.

bogdandrutu commented:
@anuraaga very nice review. thanks

shengxil commented on sdk_contrib/aws_v1_support/src/main/java/io/opentelemetry/sdk/contrib/trace/aws/propagation/AWSXRayPropagator.java:
```diff
@@ -0,0 +1,238 @@
+
+  private static final Logger logger = Logger.getLogger(AWSXRayPropagator.class.getName());
+
+  static final String TRACE_HEADER_KEY = "X-Amzn-Trace-Id";
```
will fix in rev2

shengxil commented on sdk_contrib/aws_v1_support/src/main/java/io/opentelemetry/sdk/contrib/trace/aws/propagation/AWSXRayPropagator.java:
```diff
@@ -0,0 +1,238 @@
+ * OpenTelemetry.setPropagators(
+ *   DefaultContextPropagators.builder()
+ *     .addHttpTextFormat(new HttpTraceContext())
+ *     .addHttpTextFormat(new AWSXRayPropagator()).build());
```
will fix in rev2

jkwatson commented:
It doesn't look like this depends on the SDK at all, so it should be part of the API `contrib` hierarchy, rather than sdk_contrib.

jkwatson commented:
@shengxil The CI build is failing on the javadoc step. Can you take a look?

anuraaga commented on contrib/trace_propagators/src/main/java/io/opentelemetry/contrib/trace/propagation/AwsXRayPropagator.java:
```diff
@@ -0,0 +1,269 @@
+ *   DefaultContextPropagators
+ *     .builder()
+ *     .addHttpTextFormat(new HttpTraceContext())
+ *     .addHttpTextFormat(new AWSXRayPropagator()).build());
```
```suggestion
 *     .addHttpTextFormat(new AWSXRayPropagator())
       .build());
```

shengxil replied:
will fix

shengxil commented:
> It doesn't look like this depends on the SDK at all, so it should be part of the API `contrib` hierarchy, rather than sdk_contrib.

Thanks for clarifying this. I've moved it to `contrib` and fixed the javadoc issue

anuraaga commented on contrib/trace_propagators/src/main/java/io/opentelemetry/contrib/trace/propagation/AwsXRayPropagator.java:
```diff
@@ -0,0 +1,269 @@
+      return TraceId.getInvalid();
+    }
+
+    // Check delimiters
```
Cool

anuraaga commented on contrib/trace_propagators/src/test/java/io/opentelemetry/contrib/trace/propagation/AwsXRayPropagatorTest.java:
```diff
@@ -0,0 +1,214 @@
+
+    assertThat(carrier)
+        .containsEntry(
+            TRACE_HEADER_KEY, "Root=" + TRACE_ID_XRAY + ";Parent=" + SPAN_ID_BASE16 + ";Sampled=1");
```
Recommend instead having inline constants for different types of trace headers, e.g.

```java
TRACE_HEADER_SAMPLED = "Root=11241341413441;Parent=12312134;Sampled=1"
TRACE_HEADER_UNSAMPLED = "Root=11241341413441;Parent=12312134;Sampled=0"
```

There's some duplication, but since this is testdata, not business logic, doesn't really affect maintenance. But it has a nice benefit that anyone, including people not familiar with X-Ray, can easily see what an X-Ray header is supposed to look like.

shengxil replied:
ok that's fair. will fix

anuraaga commented on contrib/trace_propagators/src/test/java/io/opentelemetry/contrib/trace/propagation/AwsXRayPropagatorTest.java:
```diff
@@ -0,0 +1,214 @@
+  }
+
+  @Test
+  public void extract_NotSampledContext() {
```
Couple more test cases

1) Valid header with keys in different oder
2) Valid header with additional params. We strip this right now but need to make sure this propagator can read any valid header

anuraaga replied:
Ah can we also add a test where we inject and there's some trace state? Documents that we ignore for now and will give a target to update when actually implementing it.

shengxil replied:
Sure, will add those tests

anuraaga commented on contrib/trace_propagators/src/main/java/io/opentelemetry/contrib/trace/propagation/AwsXRayPropagator.java:
```diff
@@ -0,0 +1,269 @@
+      return TraceFlags.builder().setIsSampled(false).build();
+    } else {
+      try {
+        return TraceFlags.fromLowerBase16(xraySampledFlag, 0);
```
Isn't this branch an invalid header?

shengxil replied:
Oops it should be.

shengxil replied:
Oops it should be.

anuraaga commented on contrib/trace_propagators/src/test/java/io/opentelemetry/contrib/trace/propagation/AwsXRayPropagatorTest.java:
```diff
@@ -0,0 +1,214 @@
+    Map<String, String> invalidHeaders = new LinkedHashMap<>();
+    invalidHeaders.put(
+        TRACE_HEADER_KEY,
+        "Root=" + TRACE_ID_XRAY + ";Parent=" + SPAN_ID_BASE16 + ";Sampled=abcdefr");
```
I don't think this tests anything different from `_Size`. But having a test with one character that isn't `0` or `1` seems important

shengxil replied:
will change to "a"

anuraaga commented on contrib/trace_propagators/src/main/java/io/opentelemetry/contrib/trace/propagation/AwsXRayPropagator.java:
```diff
@@ -0,0 +1,269 @@
+  private static final char KV_DELIMITER = '=';
+
+  private static final String TRACE_ID_KEY = "Root";
+  private static final char TRACE_ID_LENGTH = 35;
```
```suggestion
  private static final int TRACE_ID_LENGTH = 35;
```

shengxil replied:
will fix

shengxil commented on sdk_contrib/aws_v1_support/src/main/java/io/opentelemetry/sdk/contrib/trace/aws/propagation/AWSXRayPropagator.java:
```diff
@@ -0,0 +1,258 @@
+    Objects.requireNonNull(setter, "setter");
+
+    Span span = TracingContextUtils.getSpanWithoutDefault(context);
+    if (span == null || !span.getContext().isValid()) {
```
will fix

jkwatson commented on contrib/trace_propagators/src/main/java/io/opentelemetry/contrib/trace/propagation/AwsXRayPropagator.java:
```diff
@@ -0,0 +1,266 @@
+    }
+
+    // Check version trace id version
+    if (xrayTraceId.charAt(0) != TRACE_ID_VERSION) {
```
If you made the TRACE_ID_VERSION a string, you could avoid the String.valueOf call above, and just use `startsWith` for this call. Might read a little cleaner in both cases.

shengxil commented:
> Just one tiny comment, but it's not big enough not to approve.
> 
> Thanks for the contribution!

Thanks for the review! Addressed it

shengxil commented:
Squashed all commits.

jkwatson commented on contrib/trace_propagators/build.gradle:
```diff
@@ -12,6 +12,9 @@ ext.moduleName = "io.opentelemetry.contrib.trace.propagation"
     api project(':opentelemetry-api')
 
+    implementation libraries.guava,
+            platform(boms.guava)
```
I just realized that this means the B3 and Jaeger propagators now depend on guava. I'm concerned that will be surprising, and potentially lead to version conflicts down the road. 

Opinions, @carlosalberto and @bogdandrutu ?

shengxil replied:
I'm actually only using the `@VisibleForTesting` annotation (which I think is valuable for b3 and jaeger propagators as well). The main package has depended on Guava too. IMO thinking long term the `contrib` will need Guava now or later. It is better to have it before GA

jkwatson replied:
My concern is that you're forcing a guava dependency on the B3 and Jaeger propagators when they don't need it (and likely don't want it). I'd rather just replace the annotation with a comment on those methods, rather than add guava here.

anuraaga replied:
We were initially working in the sdk_contrib repo and I think there it makes sense to use Guava where helpful since the sdk depends on it anyways. But agree for the API we don't really need it, at least for now.

anuraaga commented on contrib/trace_propagators/src/main/java/io/opentelemetry/contrib/trace/propagation/AwsXRayPropagator.java:
```diff
@@ -0,0 +1,266 @@
+ */
+public class AwsXRayPropagator implements HttpTextFormat {
+
+  @VisibleForTesting static final String TRACE_HEADER_KEY = "X-Amzn-Trace-Id";
```
```suggestion
  // Visible for testing
  static final String TRACE_HEADER_KEY = "X-Amzn-Trace-Id";
```

anuraaga commented on contrib/trace_propagators/src/main/java/io/opentelemetry/contrib/trace/propagation/AwsXRayPropagator.java:
```diff
@@ -0,0 +1,266 @@
+  @SuppressWarnings("StringSplitter")
+  private static <C> SpanContext getSpanContextFromHeader(C carrier, Getter<C> getter) {
+    String traceHeader = getter.get(carrier, TRACE_HEADER_KEY);
+    if (Strings.isNullOrEmpty(traceHeader)) {
```
```suggestion
    if (traceHeader == null || traceHeader.isEmpty()) {
```

For just this one line I guess we may as well inline this rather than worry about guava or helper methods etc. Alternatively, we don't really need the empty check as an empty string is invalid and will always work with our below logic but it'd be somewhat less optimal. Not sure if a `getter` will return empty strings often enough for us to worry about it though.

anuraaga commented on contrib/trace_propagators/build.gradle:
```diff
@@ -12,6 +12,9 @@ ext.moduleName = "io.opentelemetry.contrib.trace.propagation"
 dependencies {
     api project(':opentelemetry-api')
 
+    implementation libraries.guava,
+            platform(boms.guava)
+
```
```suggestion
```

anuraaga commented on contrib/trace_propagators/src/main/java/io/opentelemetry/contrib/trace/propagation/AwsXRayPropagator.java:
```diff
@@ -0,0 +1,266 @@
+
+package io.opentelemetry.contrib.trace.propagation;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
```
```suggestion
```

lupengamzn commented on contrib/trace_propagators/src/main/java/io/opentelemetry/contrib/trace/propagation/AwsXRayPropagator.java:
```diff
@@ -0,0 +1,264 @@
+
+      String value = trimmedPart.substring(equalsIndex + 1);
+
+      if (trimmedPart.startsWith(TRACE_ID_KEY)) {
```
Once the parse for each section is done, we can try to check the value and stop immediately if the value is invalid, instead of parsing all sections and then checking them one by one later. Thoughts?

lupengamzn replied:
@anuraaga
